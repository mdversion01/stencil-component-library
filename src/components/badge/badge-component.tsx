import { Component, Prop, h, Element, Event, EventEmitter } from '@stencil/core';
import { buildBadgeClass } from '../../utils/buildBadgeClass';
import { renderTokenSpan } from '../../utils/render.helpers';
import { logInfo } from '../../utils/log-debug';

@Component({
  tag: 'badge-component',
  styleUrls: ['badge.scss',],
  shadow: false,
})
export class Badge {
  @Element() el: HTMLElement;

  @Event({ bubbles: true }) customClick: EventEmitter;

  @Prop() absolute: boolean = false;
  @Prop() ariaLabelledby?: string;
  @Prop() ariaDescribedby?: string;
  @Prop() bdgPosition: string = '';
  @Prop() backgroundColor: string = '';
  @Prop() bordered: boolean = false;
  @Prop() bottom: string = '';
  @Prop() classNames: string = '';
  @Prop() color: string = '';
  @Prop() disabled: boolean = false;
  @Prop() dot: boolean = false;
  @Prop() elevation: string = '';
  @Prop() icon: boolean = false;
  @Prop() styles: string = '';
  @Prop() inlineStyles: string = '';
  @Prop() inset: boolean = false;
  @Prop() left: string = '';
  @Prop() offsetX: string = '12';
  @Prop() offsetY: string = '12';
  @Prop() outlined: boolean = false;
  @Prop() pulse: boolean = false;
  @Prop() right: string = '';
  @Prop() shape: string = '';
  @Prop() size: string = '';
  @Prop() token: boolean = false;
  @Prop() top: string = '';
  @Prop() variant: string = '';
  @Prop() zIndex: string = '';
  @Prop() label: string = 'Badge';

  @Prop() devMode: boolean = false;


  private getDynamicStyles(): { [key: string]: string } {
    const styles: { [key: string]: string } = {};

    if (this.color) styles.color = this.color;
    if (this.backgroundColor) styles.backgroundColor = this.backgroundColor;
    if (this.absolute) styles.position = 'absolute';
    if (this.left) styles.left = `${this.left}px`;
    if (this.right) styles.right = `${this.right}px`;
    if (this.top) styles.top = `${this.top}px`;
    if (this.bottom) styles.bottom = `${this.bottom}px`;
    if (this.zIndex) styles.zIndex = this.zIndex;
    if (this.inset) {
      styles.inset = `auto auto calc(100% - ${this.offsetX}px) calc(100% - ${this.offsetY}px)`;
    }

    logInfo(this.devMode, 'Badge', 'getDynamicStyles()', styles);
    return styles;
  }

  private parseInlineStyles(styles: string): { [key: string]: string } {
    const parsed = styles.split(';').reduce((acc, style) => {
      const [key, value] = style.split(':').map(s => s.trim());
      if (key && value) {
        acc[key] = value;
      }
      return acc;
    }, {} as { [key: string]: string });

    logInfo(this.devMode, 'Badge', 'parseInlineStyles()', parsed);
    return parsed;
  }

  private renderBadgeToken() {
    const tokenClass = buildBadgeClass({
      base: 'badge__token',
      elevation: this.elevation,
      outlined: this.outlined,
      variant: this.variant ? 'bg-' + this.variant : 'bg-primary',
      size: this.size,
      shape: this.shape,
      bordered: this.bordered,
      classNames: this.classNames,
    });

    const wrapperClass = buildBadgeClass({
      base: 'badge__token-wrapper',
      variant: this.variant,
      size: this.size,
      shape: this.shape,
      outlined: this.outlined,
      bordered: this.bordered,
      classNames: this.classNames,
    });

    const combinedStyles = {
      ...this.getDynamicStyles(),
      ...this.parseInlineStyles(this.inlineStyles),
    };

    logInfo(this.devMode, 'Badge', 'renderBadgeToken()', {
      tokenClass,
      wrapperClass,
      combinedStyles,
      ariaLabelledby: this.ariaLabelledby,
      ariaDescribedby: this.ariaDescribedby,
    });

    return (
      <div class={wrapperClass} aria-disabled={this.disabled ? 'true' : 'false'}>
        <slot />
        <span class="badge__wrapper">
          {renderTokenSpan({
            className: tokenClass,
            styles: combinedStyles,
            ariaLabel: this.label,
            ariaLabelledby: this.ariaLabelledby,
            ariaDescribedby: this.ariaDescribedby,
          })}
        </span>
      </div>
    );
  }

  private renderDotBadge() {
    const tokenClass = buildBadgeClass({
      base: 'badge--dot',
      elevation: this.elevation,
      variant: this.variant ? 'bg-' + this.variant : 'bg-danger',
      size: this.size,
      shape: this.shape,
      pulse: this.pulse,
      bordered: this.bordered,
      classNames: this.classNames,
    });

    const wrapperClass = buildBadgeClass({
      base: 'badge__token-wrapper',
      variant: this.variant,
      size: this.size,
      shape: this.shape,
      outlined: this.outlined,
      bordered: this.bordered,
      classNames: this.classNames,
    });

    const combinedStyles = {
      ...this.getDynamicStyles(),
      ...this.parseInlineStyles(this.inlineStyles),
    };

    return (
      <div class={wrapperClass} aria-disabled={this.disabled ? 'true' : 'false'}>
        <slot />
        <span class="badge__wrapper">
          {renderTokenSpan({
            className: tokenClass,
            styles: combinedStyles,
            ariaLabel: this.label,
            ariaLabelledby: this.ariaLabelledby,
            ariaDescribedby: this.ariaDescribedby,
          })}
        </span>
      </div>
    );
  }

  private renderPositionedBadge(classAttr: string) {
    return (
      <div
        class={`${classAttr} ${this.bdgPosition === 'left' ? 'me-1' : 'ms-1'}`}
        aria-disabled={this.disabled ? 'true' : 'false'}
        aria-hidden={!this.token && !this.dot ? 'true' : null}
        aria-labelledby={this.ariaLabelledby}
        aria-describedby={this.ariaDescribedby}
      >
        <slot />
      </div>
    );
  }

  private renderDefaultBadge(classAttr: string) {
    return (
      <div
        class={classAttr}
        aria-disabled={this.disabled ? 'true' : 'false'}
        aria-hidden={!this.token && !this.dot ? 'true' : null}
        aria-labelledby={this.ariaLabelledby}
        aria-describedby={this.ariaDescribedby}
        style={this.parseInlineStyles(this.inlineStyles)}
      >
        <slot />
        {this.icon && (
          <span class="icon">
            <slot name="icon" />
          </span>
        )}
      </div>
    );
  }

  render() {
    const classAttr = buildBadgeClass({
      base: this.token ? 'badge__badge' : 'badge',
      outlined: this.outlined,
      variant: this.backgroundColor ? '' : this.variant ? 'bg-' + this.variant : 'bg-secondary',
      backgroundColor: this.variant ? '' : this.backgroundColor,
      elevation: this.elevation,
      size: this.size,
      shape: this.shape === 'pill' ? 'rounded-pill' : this.shape === 'square' ? 'square' : '',
      bordered: this.bordered,
      classNames: this.classNames,
    });

    logInfo(this.devMode, 'Badge', 'render() decision tree', {
      token: this.token,
      dot: this.dot,
      bdgPosition: this.bdgPosition,
      classAttr,
    });

    if (this.token) return this.renderBadgeToken();
    if (this.dot) return this.renderDotBadge();
    if (this.bdgPosition) return this.renderPositionedBadge(classAttr);
    return this.renderDefaultBadge(classAttr);
  }
}
