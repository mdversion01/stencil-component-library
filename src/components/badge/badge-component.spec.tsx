import { newSpecPage } from '@stencil/core/testing';
import { Badge } from './badge-component';

describe('badge-component', () => {
  it('renders default badge', async () => {
    const page = await newSpecPage({
      components: [Badge],
      html: `<badge-component label="Default badge"></badge-component>`,
    });
    const rootDiv = page.root.querySelector('.badge') as HTMLDivElement;
    expect(rootDiv).toBeTruthy();
    expect(rootDiv.className).toContain('bg-secondary');
    expect(rootDiv.getAttribute('aria-hidden')).toBe('true');
    expect(page.root).toMatchSnapshot();
  });

  it('renders token badge with ARIA props and inlineStyles', async () => {
    const page = await newSpecPage({
      components: [Badge],
      html: `<badge-component
        token
        variant="info"
        size="lg"
        label="Token badge"
        aria-labelledby="badgeLabel"
        aria-describedby="badgeDesc"
        inline-styles="background-color: yellow; border-radius: 8px;"
      ></badge-component>`,
    });

    const token = page.root.querySelector('.badge__token') as HTMLSpanElement;
    expect(token).toBeTruthy();
    expect(token.className).toContain('bg-info');
    expect(token.className).toContain('lg');
    expect(token.getAttribute('aria-label')).toBe('Token badge');
    expect(token.getAttribute('aria-labelledby')).toBe('badgeLabel');
    expect(token.getAttribute('aria-describedby')).toBe('badgeDesc');
    const style = token.getAttribute('style') || '';
    expect(style).toContain('background-color: yellow');
    expect(style).toContain('border-radius: 8px');

    expect(page.root).toMatchSnapshot();
  });

  it('renders dot badge with shape and pulse', async () => {
    const page = await newSpecPage({
      components: [Badge],
      html: `<badge-component dot pulse shape="circle" variant="danger"></badge-component>`,
    });

    const dot = page.root.querySelector('.badge--dot') as HTMLSpanElement;
    expect(dot).toBeTruthy();
    expect(dot.className).toContain('bg-danger');
    expect(dot.className).toMatch(/pulse/);
    // shape mapping can vary; assert presence of typical rounded/shape marker
    expect(dot.className).toMatch(/circle|rounded/);

    expect(page.root).toMatchSnapshot();
  });

  it('renders positioned badge (left)', async () => {
    const page = await newSpecPage({
      components: [Badge],
      html: `<badge-component bdg-position="left" label="Positioned"></badge-component>`,
    });
    const node = page.root.querySelector('div')!;
    expect(node.className).toMatch(/me-1/);
    expect(node.getAttribute('aria-hidden')).toBe('true');
    expect(page.root).toMatchSnapshot();
  });

  it('renders positioned badge (right)', async () => {
    const page = await newSpecPage({
      components: [Badge],
      html: `<badge-component bdg-position="right" label="Positioned"></badge-component>`,
    });
    const node = page.root.querySelector('div')!;
    expect(node.className).toMatch(/ms-1/);
    expect(page.root).toMatchSnapshot();
  });

  it('applies dynamic styles correctly (token path)', async () => {
    const page = await newSpecPage({
      components: [Badge],
      html: `<badge-component
      token
      absolute
      left="10"
      top="20"
      z-index="5"
      label="Styled"
    ></badge-component>`,
    });

    const span = page.root.querySelector('.badge__token') as HTMLSpanElement;
    const style = span.getAttribute('style') || '';

    expect(style).toContain('position: absolute');
    expect(style).toContain('left: 10px');
    expect(style).toContain('top: 20px');
    expect(style).toContain('z-index: 5');
    expect(page.root).toMatchSnapshot();
  });

  // ── NEW: lock outlined + bordered + elevation (token) ─────────
  it('token badge: outlined + bordered + elevation classes present and stable', async () => {
    const page = await newSpecPage({
      components: [Badge],
      html: `<badge-component token outlined bordered elevation="3" variant="primary" label="Combo"></badge-component>`,
    });
    const token = page.root.querySelector('.badge__token') as HTMLSpanElement;
    expect(token).toBeTruthy();

    const cls = token.className;
    expect(cls).toContain('elevated-3');
    expect(cls).toMatch(/outlined|badge--outlined/); // support either naming
    expect(cls).toMatch(/bordered|badge--bordered/); // support either naming

    expect(page.root).toMatchSnapshot();
  });

  // ── NEW: lock bordered + elevation (dot) ──────────────────────
  it('dot badge: bordered + elevation classes present and stable', async () => {
    const page = await newSpecPage({
      components: [Badge],
      html: `<badge-component dot bordered elevation="2" variant="success"></badge-component>`,
    });
    const dot = page.root.querySelector('.badge--dot') as HTMLSpanElement;
    expect(dot).toBeTruthy();

    const cls = dot.className;
    expect(cls).toContain('elevated-2');
    expect(cls).toMatch(/bordered|badge--bordered/);

    expect(page.root).toMatchSnapshot();
  });
});
