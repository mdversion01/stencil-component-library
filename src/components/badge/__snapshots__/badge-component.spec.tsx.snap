// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`badge-component applies dynamic styles correctly (token path) 1`] = `
<badge-component absolute="" label="Styled" left="10" token="" top="20" z-index="5">
  <!---->
  <div aria-disabled="false" class="badge__token-wrapper">
    <span class="badge__wrapper">
      <span aria-atomic="true" aria-label="Styled" aria-live="polite" class="badge__token bg-primary" role="status" tabindex="0" style="position: absolute; left: 10px; top: 20px; z-index: 5;"></span>
    </span>
  </div>
</badge-component>
`;

exports[`badge-component dot badge: bordered + elevation classes present and stable 1`] = `
<badge-component bordered="" dot="" elevation="2" variant="success">
  <!---->
  <div aria-disabled="false" class="badge__token-wrapper success token-bordered">
    <span class="badge__wrapper">
      <span aria-atomic="true" aria-label="Badge" aria-live="polite" class="badge--dot bg-success elevated-2 token-bordered" role="status" tabindex="0"></span>
    </span>
  </div>
</badge-component>
`;

exports[`badge-component renders default badge 1`] = `
<badge-component label="Default badge">
  <!---->
  <div aria-disabled="false" aria-hidden="true" class="badge bg-secondary"></div>
</badge-component>
`;

exports[`badge-component renders dot badge with shape and pulse 1`] = `
<badge-component dot="" pulse="" shape="circle" variant="danger">
  <!---->
  <div aria-disabled="false" class="badge__token-wrapper circle danger">
    <span class="badge__wrapper">
      <span aria-atomic="true" aria-label="Badge" aria-live="polite" class="badge--dot bg-danger circle pulse" role="status" tabindex="0"></span>
    </span>
  </div>
</badge-component>
`;

exports[`badge-component renders positioned badge (left) 1`] = `
<badge-component bdg-position="left" label="Positioned">
  <!---->
  <div aria-disabled="false" aria-hidden="true" class="badge bg-secondary me-1"></div>
</badge-component>
`;

exports[`badge-component renders positioned badge (right) 1`] = `
<badge-component bdg-position="right" label="Positioned">
  <!---->
  <div aria-disabled="false" aria-hidden="true" class="badge bg-secondary ms-1"></div>
</badge-component>
`;

exports[`badge-component renders token badge with ARIA props and inlineStyles 1`] = `
<badge-component aria-describedby="badgeDesc" aria-labelledby="badgeLabel" inline-styles="background-color: yellow; border-radius: 8px;" label="Token badge" size="lg" token="" variant="info">
  <!---->
  <div aria-disabled="false" class="badge__token-wrapper info lg">
    <span class="badge__wrapper">
      <span aria-atomic="true" aria-describedby="badgeDesc" aria-label="Token badge" aria-labelledby="badgeLabel" aria-live="polite" class="badge__token bg-info lg" role="status" tabindex="0" style="background-color: yellow; border-radius: 8px;"></span>
    </span>
  </div>
</badge-component>
`;

exports[`badge-component token badge: outlined + bordered + elevation classes present and stable 1`] = `
<badge-component bordered="" elevation="3" label="Combo" outlined="" token="" variant="primary">
  <!---->
  <div aria-disabled="false" class="badge__token-wrapper outlined primary token-bordered">
    <span class="badge__wrapper">
      <span aria-atomic="true" aria-label="Combo" aria-live="polite" class="badge__token bg-primary elevated-3 outlined token-bordered" role="status" tabindex="0"></span>
    </span>
  </div>
</badge-component>
`;
